### DISCUSSION TITLE
Some binary files are inconsistent between different compiles

### DISCUSSION THREAD

**Discussion by EricLin89** [rabbitmq/rabbitmq-server#3431] [Other] [answered]
We are using RabbitMQ in a commercial product, which delivers the installation package(with RabbitMQ included) to customer. 
So it requires all the binary files in the package are consistent between different compiles, under a given release of source code.

We are compiling RabbitMQ 3.8.14 with erlang 23.0 and Elixir 1.10.4. The compilation process is run inside a chroot environment, which was built by running "yum install basesystem erlang elixir ... ", the yum repos are fixed.

We have a tool for comparing 2 compilation results, and it complains about some binary files are inconsistent. Those are:
Elixir.Collectable.beam
Elixir.CSV.Encode.beam
Elixir.Enumerable.beam
Elixir.IEx.Info.beam
Elixir.Inspect.beam
Elixir.JSON.Decoder.beam
Elixir.JSON.Encoder.beam
Elixir.List.Chars.beam
Elixir.RabbitMQ.CLI.Core.DataCoercion.beam
Elixir.String.Chars.beam
The above files are extracted from "rabbitmqctl" file.

I've made some digging into the makefiles, it seems the make process compiles something from elixir into rabbitmqctl using the "mix" command. But what happens inside "mix" is just beyond my capability.

Another point is this inconsistency only happens when the two compiles are executed on different machines, which is weired because the chroot environment should block the difference of hostos. Also actually the different machines are installed with the same host images.

Acually we have some tools to fix some variables(timestamps, randoms, etc) through the compilation process, but I just don't know where to "fix". Could anyone help to figure it out?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
`mix` is the Elixir build tool. It is not a part of RabbitMQ.

One possible reason can be that the [`+deterministic`](https://erlang.org/doc/man/compile.html) Erlang compiler flag or its Elixir compiler equivalent is not used, so environment-specific metadata can make it to the `.beam` files.

@pjk25 may have more to add.

**Reply by EricLin89** [reply]
thanks, I'll also post it to Elixir community.

**Reply by HoloRin** [reply]
Currently `+deterministic` is only used in the bazel compilation path, so if you are building with Make, it won't apply. It was added for CI efficiency rather a general goal of deterministic builds. However, I would like to minimize the differences between the results of Make and Bazel, so it is probably worth adding the flag to the Makefile side of things. It seems there is enough consistency in your build env already where even without that flag, files from the Erlang side are consistent.

Only the rabbitmq cli contains Elixir code, and as Michael mentions, `mix` is the tool used to compile Elixir. Both Make and Bazel invoke mix to compile the cli code. I don't know off hand if mix as the equivalent of a `+deterministic` flag. The files you mention would have in fact been produced by that part of the toolchain, so I will be curious to hear if the Elixir community has any specific guidance about making compilation more consistent.

**Reply by HoloRin** [reply]
I found this mention of `+deterministic` with elixir https://github.com/elixir-lang/elixir/issues/8986 but have not yet had a chance to try it out

**Reply by EricLin89** [reply]
Thanks, I'll give it a try.
Also from the elixir community, I was instructed to check the module_info(), which comes out the "vsn" attributes are different. I'll see what Jose is going to say. https://github.com/elixir-lang/elixir/issues/11249

**Reply by HoloRin** [reply]
https://github.com/rabbitmq/rabbitmq-server/pull/3440

**Reply by HoloRin** [reply]
#3442

**Reply by EricLin89** [reply]
I tried to adopt the changes of `deps/rabbitmq_cli/Makefile` from https://github.com/rabbitmq/rabbitmq-server/pull/3442. Unfortunately, it didn't work out for me.

**Comment by EricLin89** [comment, marked as answer]
This issue is solved by https://github.com/elixir-lang/elixir/issues/11249
Thanks for the helpful discussions.
